Create v1 HPA types

piosz opened the pull request:
This PR creates a new api group `autoscaling`, moves HPA types there with support for `extensions/v1beta1`.

This is still in progress (unit&integration tests should be fixed) although I'm sending for early feedback.

In a follow up PR I'll make changes in HPA types described in https://github.com/kubernetes/kubernetes/issues/18527#issuecomment-173220603 and https://github.com/kubernetes/kubernetes/issues/18528#issuecomment-172985932. I don't want to complicate this PR even more. Hopefully it's ok to make a breaking changes shortly after introducing this version.

piosz commented:
cc @bgrant0607 @davidopp @deads2k @fgrzadkowski @liggitt @mwielgus @ncdc @nikhiljindal @smarterclayton
@kubernetes/goog-csi @kubernetes/autoscaling

piosz commented:
This PR basically implements what @smarterclayton proposed in https://github.com/kubernetes/kubernetes/issues/18528#issuecomment-176544907

lavalamp commented:
@nikhiljindal can you review this one?

davidopp commented:
cc/ @madhusudancs

piosz commented:
FYI @madhusudancs this PR creates autoscaling group.

nikhiljindal commented on pkg/client/unversioned/autoscaling.go:
```diff
@@ -0,0 +1,82 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
```
Need to change to 2016 everywhere?

piosz replied:
This is checked by verify-boilerplate script. We should update the script i change everywhere I guess.

lavalamp replied:
Only change in new files.

piosz replied:
Fixed

nikhiljindal commented on pkg/apis/autoscaling/v1/types.go:
```diff
@@ -0,0 +1,873 @@
+
+// A ThirdPartyResource is a generic representation of a resource, it is used by add-ons and plugins to add new resource
+// types to the API.  It consists of one or more Versions of the api.
+type ThirdPartyResource struct {
```
Do you need all these resources here? (third party, ingress, deployments, etc)

piosz replied:
Definitely not. I haven't included them. Probably were added during rebase.

nikhiljindal commented on pkg/apis/autoscaling/register.go:
```diff
@@ -0,0 +1,55 @@
+
+// Adds the list of known types to api.Scheme.
+func addKnownTypes(scheme *runtime.Scheme) {
+	// TODO this gets cleaned up when the types are fixed
```
Can you please explain this TODO more?
sorry, its not clear to me what you want to say here.

piosz replied:
Copy pasted from extensions/register.go. I'll remove it.

nikhiljindal commented on pkg/apis/autoscaling/v1/conversion_generated.go:
```diff
@@ -0,0 +1,438 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
```
Was this file autogenerated with 2015?

piosz replied:
Yes.

nikhiljindal commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -267,6 +268,23 @@ func Run(s *options.APIServer) error {
 	storageDestinations.AddAPIGroup("", etcdStorage)
 
+	// autoscaling group
+	if !apiGroupVersionOverrides["autoscaling/v1b"].Disable {
```
autoscaling/v1b->autoscaling/v1 (remove extra 'b' at the end)

piosz replied:
Done.

nikhiljindal commented:
This PR is huge.
Would be great if you can break it down. The last 2 commits can surely go in a separate PR.

Re: making a breaking change to v1: yes its fine (thats what we did when we moved from v1beta3 to v1, we first copied types as is and then made changes). We just need to make sure that there is no release in between.

piosz commented:
@nikhiljindal I tried to break the PR down unfortunately I got into some issues and I didn't have enough time to do it. Anyway the PR is split across a bunch of commits so it should be reviewable.

madhusudancs commented:
@piosz What exactly are you going to implement in this -  https://github.com/kubernetes/kubernetes/issues/18528#issuecomment-172985932? Just want to draw your attention to PR #19802, commit https://github.com/madhusudancs/kubernetes/commit/aebc0162f65aeeb68af8638859cb97b12cb5b57d which partly implements what is described in that comment. Specifically it implements what is described here - https://github.com/kubernetes/kubernetes/issues/18528#issuecomment-176995917 and waiting on autoscaling API group to be created. However, it doesn't implement the comment you linked to fully. We need to collaborate on this to avoid effort duplication.

Btw, these PR is blocking a bunch of other PRs that are already LGTM'ed.

piosz commented:
@madhusudancs this PR basically adds a new API group `autoscaling`. In a follow up PR I'm going to apply 2 changes to `HorizontalPodAutoscaler` described in https://github.com/kubernetes/kubernetes/issues/18528#issuecomment-174524521 point 1.

I know you are working on graduating Scale subresource to GA (and making required changes there) which is not goal for this and follow up PR.

Since @madhusudancs is blocked on this PR and I'm also planning to make some changes which depends on this PR could you please @nikhiljindal @lavalamp prioritize reviewing it? Also both changes @madhusudancs and mine are blocker for 1.2 and we are pretty close to it.

madhusudancs commented:
@piosz That makes sense. Thanks for that clarification.

lavalamp commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -438,33 +456,24 @@ func parseRuntimeConfig(s *options.APIServer) (map[string]genericapiserver.APIGr
 	}
 	_ = disableLegacyAPIs // hush the compiler while we don't have legacy APIs to disable.
 
-	// "api/v1={true|false} allows users to enable/disable v1 API.
```
The changes between here until the `for key := range s.RuntimeConfig {` loop-- what is the purpose?

piosz replied:
I basically fixed todo from line 457: // TODO: Make this a loop over all group/versions when there are more of them.
and added support for `autoscaling/v1`. The logic should remain the same (except for support for the new api group).

lavalamp commented on docs/user-guide/kubectl/kubectl_autoscale.md:
```diff
@@ -64,7 +64,7 @@ $ kubectl autoscale rc foo --max=5 --cpu-percent=80
       --dry-run[=false]: If true, only print the object that would be sent, without creating it.
   -f, --filename=[]: Filename, directory, or URL to a file identifying the resource to autoscale.
-      --generator="horizontalpodautoscaler/v1beta1": The name of the API generator to use. Currently there is only 1 generator.
+      --generator="horizontalpodautoscaler/v1": The name of the API generator to use. Currently there is only 1 generator.
```
Did you intend to change the generator's version?

piosz replied:
Which should be the default one assuming we have two: v1beta1 and v1?

nikhiljindal commented on pkg/apis/autoscaling/v1/types.go:
```diff
@@ -0,0 +1,118 @@
+
+// ListOptions is the query options to a standard REST list call.
+// Copy-pasted from api/v1.
+type ListOptions struct {
```
I am not sure if we really need a copy.
Can we reuse v1.ListOptions?
I see that we do that for DeleteOptions in extensions/v1beta: https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/extensions/v1beta1/register.go#L60

cc @smarterclayton @lavalamp @caesarxuchao

piosz replied:
Change to v1.ListOptions, seems to work.

liggitt replied:
v1 or unversioned? I thought we were trying to change dependencies on v1 utility objects like this. @lavalamp, @deads2k, does using v1 here seem right to you?

smarterclayton replied:
No, we are not using unversioned for these objects anymore.  We are using
versioned objects, and we are not moving things into unversioned/ for
params for things like this yet.

On Thu, Feb 4, 2016 at 8:28 AM, Jordan Liggitt <notifications@github.com>
wrote:

> In pkg/apis/autoscaling/v1/types.go
> <https://github.com/kubernetes/kubernetes/pull/20501#discussion_r51871188>
> :
>
> > +	Status HorizontalPodAutoscalerStatus `json:"status,omitempty"`
> > +}
> > +
> > +// list of horizontal pod autoscaler objects.
> > +type HorizontalPodAutoscalerList struct {
> > +	unversioned.TypeMeta `json:",inline"`
> > +	// Standard list metadata.
> > +	unversioned.ListMeta `json:"metadata,omitempty"`
> > +
> > +	// list of horizontal pod autoscaler objects.
> > +	Items []HorizontalPodAutoscaler `json:"items"`
> > +}
> > +
> > +// ListOptions is the query options to a standard REST list call.
> > +// Copy-pasted from api/v1.
> > +type ListOptions struct {
>
> v1 or unversioned? I thought we were trying to change dependencies on v1
> utility objects like this. @lavalamp <https://github.com/lavalamp>,
> @deads2k <https://github.com/deads2k>, does using v1 here seem right to
> you?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20501/files#r51871188>.
>

nikhiljindal commented on pkg/apis/autoscaling/validation/validation.go:
```diff
@@ -0,0 +1,94 @@
+// ValidateHorizontalPodAutoscaler can be used to check whether the given autoscaler name is valid.
+// Prefix indicates this name will be used as part of generation, in which case trailing dashes are allowed.
+func ValidateHorizontalPodAutoscalerName(name string, prefix bool) (bool, string) {
+	// TODO: finally move it to pkg/api/validation and use nameIsDNSSubdomain function
```
I think you can directly use NameIsDNSSubdomain() method? We do that for daemon sets and deployment: https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/extensions/validation/validation.go#L236

piosz replied:
Done

lavalamp commented on pkg/apis/extensions/register.go:
```diff
@@ -73,24 +74,22 @@ func addKnownTypes(scheme *runtime.Scheme) {
-func (obj *Deployment) GetObjectKind() unversioned.ObjectKind                  { return &obj.TypeMeta }
-func (obj *DeploymentList) GetObjectKind() unversioned.ObjectKind              { return &obj.TypeMeta }
-func (obj *DeploymentRollback) GetObjectKind() unversioned.ObjectKind          { return &obj.TypeMeta }
-func (obj *HorizontalPodAutoscaler) GetObjectKind() unversioned.ObjectKind     { return &obj.TypeMeta }
```
What is the upgrade path for HPA objects? Are people currently using them? If so you can't just delete the old objects; both old and new objects need to work at the same time for one release to allow people to upgrade.

piosz replied:
Yes, people are currently using them. I haven't removed `HorizontalPodAutoscaler` and `HorizontalPodAutoscalerList` types from `extensions/v1beta1`. I moved internal representation from `extensions/types.go` to `autoscaling/types.go` and updated `extensions/register.go` to point to `autoscaling.HorizontalPodAutoscaler` and `&autoscaling.HorizontalPodAutoscalerList` instead.

So AFAIU currently with support for both `autoscaling/v1` and `extensions/v1beta1` users will be able to perform the migration.

lavalamp replied:
OK, that sounds good.

lavalamp commented on pkg/kubectl/cmd/util/factory.go:
```diff
@@ -123,16 +124,16 @@ type Factory struct {
+	RunPodV1GeneratorName                  = "run-pod/v1"
+	ServiceV1GeneratorName                 = "service/v1"
+	ServiceV2GeneratorName                 = "service/v2"
+	HorizontalPodAutoscalerV1GeneratorName = "horizontalpodautoscaler/v1"
```
Did you intend to change the generator name?

Shouldn't you leave the old generator in place?

bgrant0607 replied:
Yes, the old generator should remain so long as we support the beta API.

cc @janetkuo

piosz replied:
Created a new generator HorizontalPodAutoscalerV1, left the old one HorizontalPodAutoscalerV1Beta1. Which one should be a default?

nikhiljindal commented on pkg/kubectl/cmd/autoscale.go:
```diff
@@ -54,7 +54,7 @@ func NewCmdAutoscale(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 		},
 	}
 	cmdutil.AddPrinterFlags(cmd)
-	cmd.Flags().String("generator", "horizontalpodautoscaler/v1beta1", "The name of the API generator to use. Currently there is only 1 generator.")
```
Not sure if we can delete the generator.
cc @bgrant0607 to confirm.

We can let the old generator remain, which can continue to generate extensions.v1beta1.HPA and add a new generator to generate autoscaling.v1.HPA.
We can delete the v1beta1 generator when we delete extensions.v1beta.HPA.

piosz replied:
Fixed and commented https://github.com/kubernetes/kubernetes/pull/20501#discussion_r51868268

lavalamp commented on pkg/master/master.go:
```diff
@@ -570,7 +620,7 @@ func (m *Master) thirdpartyapi(group, kind, version string) *apiserver.APIGroupV
 
 // getExperimentalResources returns the resources for extenstions api
-func (m *Master) getExtensionResources(c *Config) map[string]rest.Storage {
+func (m *Master) getExtensionResources(c *Config, autoscalerStorage, autoscalerStatusStorage rest.Storage) map[string]rest.Storage {
```
This function is going to grow a ton of parameters if we do it this way. Maybe add them to the config object? Or pass a map[string]map[string]rest.Storage (group -> path -> rest.Storage)?

smarterclayton replied:
Or maybe create a struct to hold them and call it ExtensionStorage

piosz replied:
How about delegate this to the person who will add next parameters? The change is actually temporary - I expect to remove those two parameters for 1.3 when we will stop supporting for autoscaling objects in extensions/v1beta1.

lavalamp replied:
That's fine as long as you add a TODO with instructions to the next person.

piosz replied:
Added TODO

nikhiljindal commented on pkg/kubectl/cmd/util/factory.go:
```diff
@@ -614,6 +617,12 @@ func (c *clientSwaggerSchema) ValidateBytes(data []byte) error {
 	}
+	if gvk.Group == autoscaling.GroupName {
+		if c.c.AutoscalingClient == nil {
+			return errors.New("unable to validate: no experimental client")
```
I guess this should be "no autoscaling client" ?

piosz replied:
Fixed

lavalamp commented on pkg/master/master.go:
```diff
@@ -588,8 +638,8 @@ func (m *Master) getExtensionResources(c *Config) map[string]rest.Storage {
+	// Install only when appriopriate storages are provided.
+	if isEnabled("horizontalpodautoscalers") && autoscalerStorage != nil && autoscalerStatusStorage != nil {
 		storage["horizontalpodautoscalers"] = autoscalerStorage
 		storage["horizontalpodautoscalers/status"] = autoscalerStatusStorage
```
Are you making the extensions/hpa resource return objects of type autoscaling/hpa? Because that will break old clients.

piosz replied:
Not sure what do you mean. If you are afraid that querying api server for extensions/v1beta1/hpa will return object in version autoscaling/v1 this is not true. Also I checked manually it:
```
curl localhost:8080/apis/extensions/v1beta1/namespaces/default/horizontalpodautoscalers/myrc
{
  "kind": "HorizontalPodAutoscaler",
  "apiVersion": "extensions/v1beta1",
...

curl localhost:8080/apis/autoscaling/v1/namespaces/default/horizontalpodautoscalers/myrc
{
  "kind": "HorizontalPodAutoscaler",
  "apiVersion": "autoscaling/v1",
...
```

lavalamp replied:
Great, that looks right.

nikhiljindal commented:
I see that you have moved everything over to the new autoscaling group.
Is there a test to verify that `extensions/v1beta1/horizontalpodautoscalers` continues to work?

nikhiljindal commented:
There are lots of changes here. I have added a few comments and I see @lavalamp has also added some.

Maybe the best way to unblock @madhusudancs is to break out the changes he wants into a separate PR (I guess he only wants the autoscaling group to exist?)

madhusudancs commented:
@piosz @nikhiljindal
@bgrant0607 suggested that I should go ahead and merge my PR, because my PR is blocking a series of other PRs. So what I am going to do is merge my PR and all the PRs that depend on it. And once the PR that creates the autoscaling group is merged, I will make the required Scale type changes.

piosz commented:
Comments applied, rebased. PTAL

piosz commented:
@madhusudancs SGTM

piosz commented:
@k8s-bot test this please

lavalamp commented:
Just one nit from me (add TODO). Maybe @nikhiljindal will see something else.

piosz commented:
Added integration test for the change in separate commit.

piosz commented:
Thanks for the review @lavalamp. Applied you comment.

piosz commented:
@k8s-bot test this issue: #20655

nikhiljindal commented:
@erictune pointed out that we cannot upgrade the controller to use v1 API in 1.2, since the controller breaks if it is upgraded before api server (or if api server is rolled back)

We should let the controller talk v1beta1 to apiserver (which will support both v1 and v1beta1) in 1.2
In 1.3, apiserver will continue to support both v1 and v1beta1, but we will move controller to v1.
In 1.4, apiserver will stop supporting v1beta1.

nikhiljindal commented on pkg/controller/namespace/namespace_controller.go:
```diff
@@ -468,13 +477,13 @@ func deletePersistentVolumeClaims(kubeClient clientset.Interface, ns string) err
 	return nil
 }
 
-func deleteHorizontalPodAutoscalers(expClient extensions_unversioned.ExtensionsInterface, ns string) error {
```
Controller should continue to use extensions client.
We will switch it to autoscaling client in 1.3

piosz replied:
Not changing as explained in https://github.com/kubernetes/kubernetes/pull/20501#issuecomment-180260757

nikhiljindal commented on pkg/controller/podautoscaler/horizontal.go:
```diff
@@ -45,7 +47,7 @@ const (
 type HorizontalController struct {
 	scaleNamespacer unversioned_extensions.ScalesGetter
-	hpaNamespacer   unversioned_extensions.HorizontalPodAutoscalersGetter
+	hpaNamespacer   unversioned_autoscaling.HorizontalPodAutoscalersGetter
```
Continue to use extensions client. Will move to autoscaling in 1.3

piosz replied:
Not changing as explained in #20501 (comment)

nikhiljindal commented on api/swagger-spec/resourceListing.json:
```diff
@@ -14,6 +14,14 @@
     "description": "get available API versions"
    },
    {
+    "path": "/apis/autoscaling/v1",
```
There should be swagger spec files for these new paths.

piosz replied:
Isn't this file autogenerated? Have I missed some configuration changes?

nikhiljindal commented on pkg/apis/extensions/install/install_test.go:
```diff
@@ -78,10 +78,9 @@ func TestInterfacesFor(t *testing.T) {
 	daemonSetGVK := gv.WithKind("DaemonSet")
 
-	if gvk, err := registered.GroupOrDie(extensions.GroupName).RESTMapper.KindFor(gv.WithResource("horizontalpodautoscalers")); err != nil || gvk != hpaGVK {
+	if gvk, err := registered.GroupOrDie(extensions.GroupName).RESTMapper.KindFor(gv.WithResource("daemonsets")); err != nil || gvk != daemonSetGVK {
```
Why this change?

This test should continue to pass without any change right?
Since extensions/v1beta1 is expected to continue to work

piosz replied:
Good point. Reverted. The test is passing.

nikhiljindal commented on pkg/client/typed/generated/extensions/unversioned/extensions_client.go:
```diff
@@ -46,10 +45,6 @@ func (c *ExtensionsClient) Deployments(namespace string) DeploymentInterface {
 	return newDeployments(c, namespace)
 }
 
-func (c *ExtensionsClient) HorizontalPodAutoscalers(namespace string) HorizontalPodAutoscalerInterface {
```
Am not sure why are these being deleted.
Extensions client should also continue to work.
Extensions client will talk to apiserver on /extensions/v1beta1/hpa.
The new autoscaling client will talk to apiserver on /autoscaling/v1/hpa.
Both can coexist.

piosz replied:
Not changing as explained in https://github.com/kubernetes/kubernetes/pull/20501#issuecomment-180260757

nikhiljindal commented on pkg/client/unversioned/extensions.go:
```diff
@@ -44,10 +43,6 @@ type ExtensionsClient struct {
 	*RESTClient
 }
 
-func (c *ExtensionsClient) HorizontalPodAutoscalers(namespace string) HorizontalPodAutoscalerInterface {
```
Same. Dont delete this

piosz replied:
Not changing as explained in https://github.com/kubernetes/kubernetes/pull/20501#issuecomment-180260757

nikhiljindal commented on pkg/client/unversioned/testclient/testclient.go:
```diff
@@ -313,10 +330,6 @@ func (c *FakeExperimental) DaemonSets(namespace string) client.DaemonSetInterfac
 	return &FakeDaemonSets{Fake: c, Namespace: namespace}
 }
 
-func (c *FakeExperimental) HorizontalPodAutoscalers(namespace string) client.HorizontalPodAutoscalerInterface {
```
Dont delete this

piosz replied:
Not changing as explained in https://github.com/kubernetes/kubernetes/pull/20501#issuecomment-180260757

nikhiljindal commented on pkg/controller/namespace/namespace_controller_test.go:
```diff
@@ -146,9 +152,21 @@ func testSyncNamespaceThatIsTerminating(t *testing.T, versions *unversioned.APIV
+		}
 		if containsVersion(versions, "extensions/v1beta1") {
 			resources := []unversioned.APIResource{}
-			for _, resource := range []string{"daemonsets", "deployments", "jobs", "horizontalpodautoscalers", "ingresses"} {
```
I think we can let hpa continue to be here

piosz replied:
I'd like to remove support for HPA extensions in 1.2 to not encourage users to use extensions version instead of autoscaling. That's why I removed support for it in extensions client. As I described I'd like to still support for `extensions/v1beta1` on api server level but I'd like to not use this version internally anymore.

liggitt replied:
What about things building on top of the client library that still need to talk to 1.1 servers?

piosz replied:
So they should use client library in version 1.1.

liggitt replied:
No… things like kubectl that need to talk to both new and old servers.

piosz replied:
Replied in https://github.com/kubernetes/kubernetes/pull/20501/files#r52013595

nikhiljindal commented on pkg/kubectl/describe.go:
```diff
@@ -1447,7 +1447,7 @@ type HorizontalPodAutoscalerDescriber struct {
 }
 
 func (d *HorizontalPodAutoscalerDescriber) Describe(namespace, name string) (string, error) {
-	hpa, err := d.client.Extensions().HorizontalPodAutoscalers(namespace).Get(name)
```
Continue to use extensions client?

piosz replied:
Not changing as explained in https://github.com/kubernetes/kubernetes/pull/20501#issuecomment-180260757

liggitt replied:
Doesn't this prevent kubectl from working against older servers?

piosz replied:
It's interesting question. @bgrant0607 do we have any policy here?
It's not negligible that API server should support previous versions of api for some time to allow users migration. The question is whether API server should support further versions of api which is of course impossible but this is actually the case.

If apiserver 1.1 should work with kubectl 1.2 should also apiserver 1.1 work with kubelet 1.2?

liggitt replied:
API server 1.2 supports kubectl 1.1 by not removing the endpoints. 
Kubectl 1.2 supports API server 1.1 by doing version negotiation and using 1.1 endpoints/versions/groups.

piosz replied:
Does the mechanism already exist?

nikhiljindal commented on test/e2e/horizontal_pod_autoscaling.go:
```diff
@@ -125,22 +125,22 @@ func scaleDown(name, kind string, rc *ResourceConsumer, f *Framework) {
 		},
 	}
-	_, errHPA := rc.framework.Client.Extensions().HorizontalPodAutoscalers(rc.framework.Namespace.Name).Create(hpa)
+	_, errHPA := rc.framework.Client.Autoscaling().HorizontalPodAutoscalers(rc.framework.Namespace.Name).Create(hpa)
```
Maybe add 2 tests?
One using Extensions client and one using autoscaling one?

nikhiljindal replied:
ok I see the integration test now.
That should be enough.
Ignore this comment

nikhiljindal commented:
Thanks for adding the master integration tests!

Would be great if you can add a command at https://github.com/kubernetes/kubernetes/blob/master/hack/test-cmd.sh#L1064
to make sure that `kubectl get extensions/hpa` also works.

bgrant0607 commented on pkg/apis/autoscaling/v1/types.go:
```diff
@@ -0,0 +1,118 @@
+	// API version of the referent
+	APIVersion string `json:"apiVersion,omitempty"`
+	// Subresource name of the referent
+	Subresource string `json:"subresource,omitempty"`
```
This is not what we previously agreed.

bgrant0607 commented:
The title of this PR is misleading.

bgrant0607 commented:
Changed the title.

Haven't looked in detail, but please don't enable the API by default until after all the incompatible changes are made to the API.

lavalamp commented on pkg/master/master.go:
```diff
@@ -197,12 +198,60 @@ func (m *Master) InstallAPIs(c *Config) {
+	// Install autoscaling unless disabled.
+	var autoscalerStorage, autoscalerStatusStorage rest.Storage
+	if !m.ApiGroupVersionOverrides["autoscaling/v1"].Disable {
+		autoscalerStorage, autoscalerStatusStorage = horizontalpodautoscaleretcd.NewREST(
```
It would be nice to put this in a separate function, I think the other groups all do that.

piosz replied:
Done.

piosz commented:
@bgrant0607 so I'll add all api changes in this PR as a separate commit. I know the PR will became even bigger but instead we will need 3 PRs and it takes ages to review and merged them in right order.

piosz commented:
@nikhiljindal @erictune 
>@erictune pointed out that we cannot upgrade the controller to use v1 API in 1.2, since the controller breaks if it is upgraded before api server (or if api server is rolled back)

I'm not sure whether we want to support backward compatibility between our master components (in this particular case apiserver and controller manager). As for now all components are upgraded together (there is no option to upgrade only controller manager) and apiserver is upgraded in the first step. Also looking back to the past it's seems we didn't support use case raised by @erictune. For example apiserver in 1.0 doesn't support extensions api group while some controllers in 1.1 require it.

My proposal is to support in 1.2 both v1 and v1beta1 and migrate all controllers to sue the new api, and then get rid off v1beta1 in 1.3 and this is what this PR implements.

Also @fgrzadkowski has some opinion here.

fgrzadkowski commented:
@nikhiljindal @erictune I agree that in theory this might be a problem, but in practice we control both the controllers and the upgrade process. Currently there is no way to upgrade the controller manager first and later the master. The same applies for the rollback. I'd vote for having a simpler solution here as long as we can have it. WDYT?

piosz commented:
@nikhiljindal 
>Would be great if you can add a command at https://github.com/kubernetes/kubernetes/blob/master/hack/test-cmd.sh#L1064
to make sure that kubectl get extensions/hpa and kubectl get autoscaling/hpa also work.

No sure whether it should work. In HEAD kubectl get extensions/hpa doesn't work.

piosz commented:
Comments addressed. I'm going to add required changes in API in this PR as a separate commit.

liggitt commented:
> My proposal is to support in 1.2 both v1 and v1beta1 and migrate all controllers to sue the new api, and then get rid off v1beta1 in 1.3 and this is what this PR implements.

Controllers aren't the only thing using the API. Agree on migrating controllers to the v1 API by 1.3, but not necessarily about removing v1beta1… need to decide how long to hold support for that.

piosz commented:
@liggitt are you suggesting to maintain backward compatibility with v1beta1 for longer than one release? I'm fine with that. The only thing I'm against here is to not move core Kubernetes components (in fact HPA controller and namespace controller) to use the new API.

piosz commented:
Addressed comments.

Added commit which applies changes described in https://github.com/kubernetes/kubernetes/issues/18527#issuecomment-173220603 and https://github.com/kubernetes/kubernetes/issues/18528#issuecomment-172985932

@bgrant0607 could you please review https://github.com/kubernetes/kubernetes/commit/424c2c5e23dc86a42eb2885d543fb53b1d28b752

piosz commented:
Still open questions are:
- whether we want to migrate controllers to the new api
- whether we want to migrate kubectl to the new api

nikhiljindal commented:
Re the open questions, I discussed them with @bgrant0607 and this is what we think:
* whether we want to migrate controllers to the new api: No. We upgrade and rollback controller and master together, but I dont think redhat does that. We cannot assume that everyone does that.
* whether we want to migrate kubectl to the new api: Yes. `kubectl get hpa` should request `/autoscaling/v1/hpa` on server. But `kubectl get hpa --api-version=extensions/v1beta1` should call `/extensions/v1beta1/hpa`. And `kubectl get hpa -o yaml --output-version=extensions/v1beta1` should output extensions.v1beta1 object.

piosz commented:
@nikhiljindal thanks for the update. SGTM. My proposal is that I'll revert the changes which touches controllers and kubectl, then in a follow up PR I'll update kubectl. WDYT?

nikhiljindal commented:
Sure. that sounds good. So this PR will only have master related changes, right?

And I havent looked at the recent changes, but @bgrant0607 also commented above that the new autoscaling/v1 api should be disabled by default, until all the changes are done.

piosz commented:
This PR will have new api group + changes in types definition agreed in #18528 + master changes + support for new client, so actually everything except kubectl changes. So that the api group will be enabled by default.

I'll let you know once the PR will be ready for the next round of review.

piosz commented:
As agreed with @lavalamp and @caesarxuchao I'll move back internal representation of HPA related typed from `autoscaling/types.go` to `extensions/types.go`.

piosz committed:
Created autoscaling API group and copied HPA there

piosz committed:
API changes in HPA required for going to GA

piosz committed:
Autogenerated conversions, deep copies, types

piosz committed:
Fixed and added tests

bgrant0607 commented on pkg/apis/autoscaling/v1/types.go:
```diff
@@ -0,0 +1,88 @@
+	// Kind of the referent; More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#types-kinds"
+	Kind string `json:"kind,omitempty"`
+	// Name of the referent; More info: http://releases.k8s.io/HEAD/docs/user-guide/identifiers.md#names
+	Name string `json:"name,omitempty"`
```
Is there any case where the name would not be required?

piosz replied:
I think no. Removing omitempty.

bgrant0607 commented on pkg/apis/autoscaling/v1/types.go:
```diff
@@ -0,0 +1,88 @@
+// CrossVersionObjectReference contains enough information to let you identify the referred resource.
+type CrossVersionObjectReference struct {
+	// Kind of the referent; More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#types-kinds"
+	Kind string `json:"kind,omitempty"`
```
Is there any case where the kind would not be required?

piosz replied:
I think no. Removing omitempty.

bgrant0607 commented:
Other than possibly some omitempty tag changes, the v1/types.go changes LGTM.

piosz commented:
@bgrant0607 thanks for the review!

piosz committed:
Implemented Autoscaling client

piosz commented:
All comments applied. @nikhiljindal PTAL

In the current version as agreed previously:
- the new API group `autoscaling` is created and enabled
- all kubernetes components except kubectl still use `extensions/v1beta1`
- kubectl use a random one (see #19941)

The current plan for HPA GA is:
- in 1.2 support for `autoscaling/v1` is added
- in 1.3 all kubernetes components are migrated to use `autoscaling/v1` and `extensions\v1beta1` is deprecated
- in 1.4 support for `extensions\v1beta1` is removed

lavalamp commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -279,6 +280,23 @@ func Run(s *options.APIServer) error {
+		if _, found := storageVersions[autoGroup.GroupVersion.Group]; !found {
+			glog.Fatalf("Couldn't find the storage version for group: %q in storageVersions: %v", autoGroup.GroupVersion.Group, storageVersions)
+		}
+		autoscalingEtcdStorage, err := newEtcd(s.EtcdServerList, api.Codecs, storageVersions[autoGroup.GroupVersion.Group], s.EtcdPathPrefix, s.EtcdQuorumRead)
```
I'm pretty sure there's no way to switch the storage version between extensions/v1beta1 and autoscaling/v1. I'm working on a fix in my PR.

liggitt replied:
will need good backwards compat tests to make sure previously persisted v1beta1 objects don't disappear

piosz replied:
IIUC @lavalamp is working on this in https://github.com/lavalamp/kubernetes/commit/29cd25f690a2dc40b8daae240abbc1df6550f523

nikhiljindal commented on api/swagger-spec/resourceListing.json:
```diff
@@ -24,6 +24,14 @@
     "description": "get information of a group"
+   },
+   {
+    "path": "/apis/autoscaling/v1",
```
I think i had pointed this before, but cant find that comment now.
You need to add files for these new paths.

You need to fix hack/update-swagger-spec to curl these new paths and store the files.
And then fix verify-swagger-spec as well.

piosz replied:
Fixed. I remember your comment. I ask there what should I do and you explained above:)

nikhiljindal commented on pkg/client/unversioned/helper.go:
```diff
@@ -155,16 +156,25 @@ func New(c *Config) (*Client, error) {
-	if _, err := registered.Group(extensions.GroupName); err != nil {
-		return &Client{RESTClient: client, ExtensionsClient: nil, DiscoveryClient: discoveryClient}, nil
+	var autoscalingClient *AutoscalingClient
+	if _, err := registered.Group(autoscaling.GroupName); err == nil {
```
Looks like you want to use registered.IsRegistered()?

piosz replied:
Done

nikhiljindal commented on pkg/client/unversioned/horizontalpodautoscaler.go:
```diff
@@ -40,12 +40,14 @@ type HorizontalPodAutoscalerInterface interface {
-	client *ExtensionsClient
+	// Use RESTClient instead of more specific to allow reusing the type in different clients.
+	// TODO(piosz): change to AutoscalingClient once completely migrated to 'autoscaling' group.
+	client *RESTClient
```
cc @caesarxuchao

liggitt replied:
what is this needed for? I'd rather take two clients in the short term than expose RESTClient here
cc @deads2k

piosz replied:
To avoid duplication of the code. Alternatively I can create horizontalPodAutoscalersV1 which has the same logic but takes AutoscalingClient or implemented the solution you proposed with 'if' in each method to check which client is not nil.

deads2k replied:
> To avoid duplication of the code. Alternatively I can create horizontalPodAutoscalersV1 which has the same logic but takes AutoscalingClient or implemented the solution you proposed with 'if' in each method to check which client is not nil.

Relaxing the constraint in this way makes it very easy to accidentally construct this using an incorrect type and not realize it until first touch at runtime.  The duplicated code that specific typing would produce is annoying, but it helps every consumer of this API use it correctly.  I'd rather have the duplication here and save everyone else from accidents.  That probably even ripples to two different implementations that satisfy the same interface to make the distinction easy to follow.

piosz replied:
Sounds reasonable. I'll add implementation of `horizontalPodAutoscalersV1`

piosz replied:
Done

caesarxuchao replied:
The latest code seems reasonable to me. Thanks.

nikhiljindal commented on pkg/client/unversioned/horizontalpodautoscaler_test.go:
```diff
@@ -162,7 +168,8 @@ func TestHorizontalPodAutoscalerWatch(t *testing.T) {
 			Query:  url.Values{"resourceVersion": []string{}}},
-		Response: simple.Response{StatusCode: 200},
+		Response:      simple.Response{StatusCode: 200},
+		ResourceGroup: extensions.GroupName,
```
Add a test with autoscaling group?

piosz replied:
Done

nikhiljindal commented on pkg/master/master.go:
```diff
@@ -255,6 +260,46 @@ func (m *Master) InstallAPIs(c *Config) {
 	}
+
+	// Install autoscaling unless disabled. Requires 'horizontalpodautoscalers' resource in extensions group to be enabled.
+	if !m.ApiGroupVersionOverrides["autoscaling/v1"].Disable && autoscalerStorage != nil && autoscalerStatusStorage != nil {
```
I think it will be better to explicitly fail if autoscaling/v1 is enabled but autoScalerStorage is nil.

piosz replied:
Done

piosz commented:
Comments addressed.

nikhiljindal commented:
Thanks @caesarxuchao for reviewing the client changes!

The changes look fine to me.
I am not sure about the kubectl behavior. Do --api-version and --output-version flags work as expected.
I dont see any changes to test-cmd.
I see you mentioned https://github.com/kubernetes/kubernetes/issues/19941.

Not sure if we want to block this PR on that and the storage version change that @lavalamp  is doing.

@lavalamp WDYT?

piosz commented:
Rebased after #20881 is merged. cc @caesarxuchao

lavalamp commented:
Thanks for doing the first passes, @nikhiljindal. My plan today is to get the corresponding Job move finished, and then very carefully make sure we've both solved the exact same set of problems in that PR and this PR.

lavalamp commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -279,6 +280,23 @@ func Run(s *options.APIServer) error {
+		if err != nil {
+			glog.Fatalf("Invalid autoscaling storage version or misconfigured etcd: %v", err)
+		}
+		storageDestinations.AddAPIGroup(autoscaling.GroupName, autoscalingEtcdStorage)
```
I am really skeptical that this creates a codec that reads & writes the correct types.

Can you add something like this change to verify? https://github.com/lavalamp/kubernetes/commit/6ecf1a33a220c3ca095acf7d5f51228a328c7cef

piosz replied:
Done. unit, integration and cmd tests are passing. I'll try to create a cluster.

piosz replied:
From apiserver.log:
```
I0212 14:48:35.237893       6 etcd.go:60] HorizontalPodAutoscaler version looks like: {"kind":"HorizontalPodAutoscaler","apiVersion":"extensions/v1beta1","metadata":{"creationTimestamp":null},"spec":{"scaleRef":{},"maxReplicas":0},"status":{"currentReplicas":0,"desiredReplicas":0}}
```

liggitt replied:
are we not backing HPA in extensions/v1beta1 and HPA in autoscaling/v1 with the same storage?

piosz commented:
@nikhiljindal regarding you comment about kubectl, currently when I'm trying to specify `--api-version` flag in kubectl I get the following error:
```
$ kubectl get hpa --api-version=autoscaling/v1
error: API version 'autoscaling/v1' is not recognized (valid values: v1)
$ kubectl get hpa --api-version=extensions/v1beta1
error: API version 'extensions/v1beta1' is not recognized (valid values: v1)
```
It seems like a bug in kubectl

deads2k commented:
`--api-version` cannot force a group, it can only specify what version to use for a given group.  If you have it actually force the group, then something like `kubectl get foo bar` (where you're getting two types) cannot reference groups in different types.

nikhiljindal commented:
I think @janetkuo has a PR to replace api-version by preferred-group-versions which will support specifying both group and version.

I am thinking it might be fine to merge this PR in its current state. Its hard to keep such a big PR in sync. Merging now will also give it time to soak and people can try it out.
We can fix any issues we discover in subsequent PRs.
Its @lavalamp's call.

piosz commented:
Rebased.

+1 to what @nikhiljindal wrote.

@lavalamp please let me know what to do with https://github.com/kubernetes/kubernetes/commit/d5d41bcfe61024776d2532af64725fda54fd9cbe

lavalamp commented:
@piosz I'm very sorry I didn't get back to you yesterday, I didn't have the job move sorted out. Now that I think the job move is mostly correct--at least server side--I think you need to do some of the things I did. Like I really don't believe that your change can work with a storage version of autoscaling/v1 without the cross-group codec stuff I did. I will try to arrange my PR so it has some commits you can cherry-pick into this one on Monday. (we have Monday off)

lavalamp committed:
changes for cross-group moves

lavalamp commented:
@piosz I think I extracted most of the things you need to add into https://github.com/kubernetes/kubernetes/commit/73fcfc2ae2057a91ac50e879c036caf73a6ca0af

but it's late on friday so I may have missed something.

Also make sure your test-go and test-integration test both ways in the KUBE_TEST_API env var (see my change for example)-- right now you only have one configuration tested. Also the update storage object test needs to be run, though maybe we can fix that up in another PR.

piosz committed:
Enabled autoscaling API group in apiserver

piosz committed:
Fixed options test

piosz commented:
@lavalamp PTAL
I cherry-picked your commit https://github.com/kubernetes/kubernetes/commit/73fcfc2ae2057a91ac50e879c036caf73a6ca0af, then made changes isomorphic with your ones. I also added appropriate test configs to unit and integration script. I'll add update storage test in a follow up PR.

piosz commented:
@k8s-bot test this issue: #IGNORE

fgrzadkowski commented:
@lavalamp Can we please have this reviewed and merged? We need to do some minor changes after this PR before Friday.

erictune commented:
@lavalamp I have made my own PR by matching what was done here.  It would be ideal for me if this could merge as is, and then any fixes done in a follow-on PR so they are isolated.  The worst thing for me would be if @piosz makes some further changes and then squashes.  Please please let's avoid that case.

lavalamp commented:
I am very sorry I did not review this yesterday, I was making slides for a talk. I'm reviewing it right now.

lavalamp commented on pkg/api/testapi/testapi.go:
```diff
@@ -69,6 +72,12 @@ func init() {
+	if _, ok := Groups[autoscaling.GroupName]; !ok {
+		Groups[autoscaling.GroupName] = TestGroup{
+			externalGroupVersion: unversioned.GroupVersion{Group: autoscaling.GroupName, Version: registered.GroupOrDie(autoscaling.GroupName).GroupVersion.Version},
+			internalGroupVersion: autoscaling.SchemeGroupVersion,
```
Shouldn't you reuse the extensions internal group here?

piosz replied:
Done

lavalamp commented on pkg/runtime/serializer/versioning/versioning.go:
```diff
@@ -132,6 +168,7 @@ func (c *codec) Decode(data []byte, defaultGVK *unversioned.GroupVersionKind, in
 		targetGV.Group = group
 		targetGV.Version = runtime.APIVersionInternal
 	} else {
+		fmt.Printf("looking for %v in %#v\n", group, c.decodeVersion)
```
This is probably from my change, but we should not log this debugging message here. Please remove?

piosz replied:
Done

lavalamp commented:
LGTM with two minor nits. Thanks for bearing with me!

@erictune I'll look at your PR next.

lavalamp commented:
This is important to get in asap, so I recommend self-applying the LGTM label after nit fixes so we can get this merged.

piosz commented:
Applied comments as a separate PR to make @erictune's life easier.

Thank you @lavalamp a lot for the review!

piosz committed:
Applied comments from the review

erictune commented:
@piosz thanks!

piosz commented:
Also in https://github.com/kubernetes/kubernetes/commit/c02e966d65febbfeafc484fdb0b45565b587120a I fixed unit tests. @lavalamp PTAL
